Generalised architecture 3 (#13)
aurora-multiphysics/cyclops: README.md

> # cyclops

Reviewer: ```suggestion
# Cyclops
```

![Overview diagram](logo.jpg "Title")

> Cyclops is a sensor placement optimisation program. Originally designed for monoblocks at the bottom of nuclear fusion reactors it searches for the optimal layout of sensors to maximise both the accuracy and reliability of the field reconstruction.

Reviewer: ```suggestion
Cyclops is a sensor suite optimisation program, designed for fusion engineering experiments.
Cyclops takes a simulated experiment and searches for the optimal layout of sensors whose readings allow for an accurate reconstruction of the full field information, while remaining robust to sensor failures.
```
Suggested change to the wording here. De-emphasising the monoblock problem at this top-level of the README and tweaking the wording a little.

Features:
1. Experiment design optimisation.
2. Machine learning tools for field reconstruction from experiment data.
3. Mesh reading tools for converting a mesh such as a MOOSE exodus file into numpy arrays.

> 

Reviewer: Consider removing this section. Reasoning: I see the functionality the basic user is interested in being the sensor suite optimisation. The field reconstruction may also be useful but also it's something we want to let the user switch in/out modularly as they need. As for the mesh reading tools, this is an "under the hood" feature which we don't need to talk about this early in the readme, since basic users don't need to know how we handle that sort of thing. As long as it's documented elsewhere, aim to keep the README concise and top-level, helping the user understand what the tool is and how to start using it.

> After reading this, go straight to '*manual/0 overview.ipynb*' to find out how to use the software. Ideally read all the manuals or you will have no idea what to do.

Reviewer: Consider moving this after the installation guide section in a new section e.g. "Getting started" pointing users to the tutorials. I think it's generally assumed that the user will have no idea what to do until they read the documentation so you don't need to state this.

The tutorials should start simple with core features and build up to more complex usage. That way, the user just needs to go through them in order until they've reached the level of complexity required for their usecase, without becoming overwhelmed. Therefore, we shouldn't usually encourage every user to read every tutorial.

Requirements: Python 3.x, Latex (Latex is recommended so the graphs are more beautiful but it is not 100% essential - see installation step 2 to get around it).

> 

Reviewer: ```suggestion
```
Can remove this as requirements are covered in installation.

# Installation

Please follow the two steps below very carefully. 

1. Install the requirements as below (note that this is for linux not windows or mac - the venv activation in setp two will be different on windows - `venv\scripts\activate`).

`python3 -m venv venv`

`source venv/bin/activate`

`pip install -r requirements.txt`

2. Install Latex on your device to let latex style graphs be plotted. 

* On MacOS I recommend MacTex https://www.tug.org/mactex/.
* On Windows I recommend MikTex https://miktex.org/.
* On Linux you can install Tex Live and other requirements with the command below.

`sudo apt-get install dvipng texlive-latex-extra texlive-fonts-recommended cm-super`

If you don't care about the latex style graphs or can't install latex then go to src/graph_management.py and then change the line in `GraphManager.__init__()` about `plt.style.use('science')` to the line below.

`plt.style.use(['science','no-latex'])`

> You are now ready to go!

Reviewer: Please write a separate installation guide for each supported OS. If an OS is supported in theory but untested (e.g. macOS?), note this. If it's unsupported, either don't mention it or say that it's unsupported. Structure should look something like this:

> # Installation
> ## Linux
> 1. Linux step 1
> 2. Linux step 2
> ## Windows
> 1. Windows step 1
> 2. Windows step 2

# Citations needed:

1. Matplotlib
2. Meshio
3. NumPy
4. SciencePlots (for Latex style graphs)
5. Scikit-learn
6. SciPy

> 

Reviewer: For a section like this, we would usually provide links to papers or even the bibtex for the dependencies which require them, rather than just the titles. Basically so that if someone publishes a paper using cyclops in the research, they can quickly find everything they need to credit in the dependency chain.

If you have time, you could try checking each of these tools' websites to see if they mention a citation they'd like us to use. Else, remove this and add it in as an issue "provide citations for dependencies" and we can handle it later when it comes to writing a paper on this stuff.

For example, here's NumPy's: https://numpy.org/citing-numpy/

# Contributers

Dom Harrington

> Luke Humphrey

Reviewer: ```suggestion
Dom Harrington, UK Atomic Energy Authority
Luke Humphrey, UK Atomic Energy Authority
Lloyd Fletcher, UK Atomic Energy Authority
```
Minor changes: one line per contributor, add institution (UKAEA), and add Lloyd as he's been closely involved
